### PR TITLE
Android Task creation wizard 

### PR DESCRIPTION
**Pull request by dmdashenkov**
This PR introduces an implementation of the task creation wizard on Android.

The wizard collects the user data and sends commands to the `TaskCreationWizard` process manager.

### CODE REVIEW

File: android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/Consumer.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.examples.todolist.lifecycle;
+
+public interface Consumer<T> {
```
Comment (armiol) [inline android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/Consumer.java]: Please document the public API.

Depending on that, I'd think whether the naming is good.

File: android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/ViewModelFactory.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import android.arch.lifecycle.ViewModel;
+import android.arch.lifecycle.ViewModelProvider.Factory;
+import android.arch.lifecycle.ViewModelProvider.NewInstanceFactory;
+import android.support.annotation.NonNull;
+
+import java.util.Map;
+
+import static com.google.common.collect.Maps.newConcurrentMap;
+
+public enum ViewModelFactory implements Factory {
```
Comment (armiol) [inline android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/ViewModelFactory.java]: Same as previously. The enum and its members have to be documented.

File: android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/Consumer.java (hunk 1 above)
Comment (armiol) [inline android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/Consumer.java]: I also don't see `@author` tags.

File: android-app/app/src/main/java/io/spine/examples/todolist/newtask/ColorPickerAdapter.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.examples.todolist.newtask;
```
Comment (armiol) [inline android-app/app/src/main/java/io/spine/examples/todolist/newtask/ColorPickerAdapter.java]: That's a strange name for a package. I'd go for `task`. And if that's not enough, I'd use `task.wizard`.
Especially given that the visibility of this class is `public`. So you don't hide anything.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.
Please also see `android-app/README.md` for the requested `README.md`.

File: android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/Consumer.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ * A functional interface representing an operation that accepts a single input argument and
+ * returns no result.
+ *
+ * <p>This interface is meant to serve for example as a callback of an asynchronous operation
+ * accepting the result of the operation.
+ *
+ * @param <T> the type of the excepted argument
+ * @author Dmytro Dashenkov
+ */
+public interface Consumer<T> {
```
Comment (armiol) [inline android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/Consumer.java]: I don't like creating interfaces like this each time...

Can we use something, that already created by someone and is available in Android? Please do a research.

File: android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/Consumer.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.examples.todolist.lifecycle;
```
Comment (armiol) [inline android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/Consumer.java]: I have some doubts as to this package name. Not sure if it exists elsewhere, but I also could not find a `package-info.java` for it — as I was hoping to read the description of the package and finally understand, whose life we are cycling.

Could you please make sure the package is described and — once you come up with the description — get back to this name. It seems way to broad to me.

File: android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/ViewModelFactory.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+import java.util.Map;
+
+import static com.google.common.collect.Maps.newConcurrentMap;
+
+/**
+ * An implementation of
+ * {@link android.arch.lifecycle.ViewModelProvider.Factory ViewModelProvider.Factory}.
+ *
+ * <p>This implementation provides two basic options for the {@link ViewModel} instantiation:
```
Comment (armiol) [inline android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/ViewModelFactory.java]: They are no "basic".

File: android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/ViewModelFactory.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+                @SuppressWarnings("unchecked") // Logically checked.
+                final T result = (T) cache.get(modelClass);
+                return result;
+            } else {
+                return ALWAYS_NEW.create(modelClass);
+            }
+        }
+    };
+
+    private static final Map<Class, ViewModel> cache = newConcurrentMap();
```
Comment (armiol) [inline android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/ViewModelFactory.java]: This looks a special documentation case to me. So can you please describe both fields? And then remove the empty line after the second one?

File: android-app/app/src/main/java/io/spine/examples/todolist/task/ColorPickerAdapter.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+            if (selector.getValue() == color) {
+                final int colorFilter = selectorView.getContext()
+                                                    .getResources()
+                                                    .getColor(R.color.colorAccentLight);
+                selectorView.setColorFilter(colorFilter);
+            } else {
+                selectorView.clearColorFilter();
+            }
+            colorView.setOnClickListener(view -> selector.setValue(color));
+
```
Comment (armiol) [inline android-app/app/src/main/java/io/spine/examples/todolist/task/ColorPickerAdapter.java]: This line is redundant.

File: android-app/app/src/main/java/io/spine/examples/todolist/task/NewTaskConfirmationFragment.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+            new SimpleDateFormat("dd MMM YYYY", Locale.getDefault());
+
+    private NewTaskViewModel model;
+
+    private TextView taskDescription;
+    private TextView taskPriority;
+    private TextView taskDueDate;
+    private RecyclerView taskLabels;
+
+    static final int POSITION_IN_WIZARD = 2;
```
Comment (armiol) [inline android-app/app/src/main/java/io/spine/examples/todolist/task/NewTaskConfirmationFragment.java]: Let's move this constant declaration up.

File: android-app/app/src/main/java/io/spine/examples/todolist/task/NewTaskConfirmationFragment.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+ *
+ * @author Dmytro Dashenkov
+ */
+public final class NewTaskConfirmationFragment extends PagerFragment {
+
+    private static final SimpleDateFormat DATE_FORMAT =
+            new SimpleDateFormat("dd MMM YYYY", Locale.getDefault());
+
+    private NewTaskViewModel model;
+
```
Comment (armiol) [inline android-app/app/src/main/java/io/spine/examples/todolist/task/NewTaskConfirmationFragment.java]: For some reason you have inserted these "section delimiter" empty lines in each of the classes like this one. However, only the author really understands the purpose. 

To me they don't look logical in every case, but get in the way of reading.

So let's do that properly and use a section separator along with the title (see core-java/wiki) or remove the empty lines at all.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

**Pull request merged**